### PR TITLE
[WIP] [docker/kubernetes]: deploy ray on kubernetes

### PR DESCRIPTION
**Pull request by knightXun**
[#2037](https://github.com/ray-project/ray/issues/2037)
I provider the simplest yaml to deploy ray.
there are many works need to do:
1. health check for ray process so that if ray process crash down, kubelet can detect it, then restart it
2. split service, we should spilt redis, scheduler, works from ray-master or ray-slave.
3. using helm
4. using dns to detect the related services
5. using persistent volume to store data
6. autoscaler 

I know few about ray, sorry for that.I will ask for your suggestions.

### CODE REVIEW

**Comment by knightXun** [comment]
redis must be spilt from ray-head

**Comment by richardliaw** [comment]
retest this please

**Comment by knightXun** [comment]
@richardliaw this is just a first edition.
I want to split redis from ray-head.I have written redis-cluster yaml.

**Comment by richardliaw** [comment]
Yep - no problem.

**Comment by robertnishihara** [comment]
@knightXun can you clarify what you mean by splitting off Redis? The Redis servers can in principle run in a different container, but using Redis cluster won't work out of the box (e.g., we use our own Redis modules). For this PR, let's avoid handling Redis in any special way.

File: docker/kubernetes/ray/ray-master.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: ray-master
+  namespace: ray-cluster
+  labels:
+    name: ray-master
+spec:
+  ports:
```
Comment (robertnishihara) [inline docker/kubernetes/ray/ray-master.yaml]: We'll need either ports open for the other Redis servers and for the object manager.

File: docker/kubernetes/ray/images/run.sh [hunk 2]
```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
+
+function run_head() {
+    ray start --head --redis-port=6379
```
Comment (robertnishihara) [inline docker/kubernetes/ray/images/run.sh]: This command starts multiple Redis servers. The ports of the other Redis servers can be specified with `--redis-shard-ports` (in a comma separated list).

This command also starts an object manager whose port can be specified with `--object-manager-port`.

File: docker/kubernetes/ray/ray-master.yaml (hunk 1 above)
Comment (knightXun) [reply docker/kubernetes/ray/ray-master.yaml]: ok

File: docker/kubernetes/ray/images/run.sh (hunk 2 above)
Comment (knightXun) [reply docker/kubernetes/ray/images/run.sh]: ok, thanks.

Comment (knightXun) [reply docker/kubernetes/ray/images/run.sh]: I will read your code and docs more carefully.

**Comment by knightXun** [comment]
@robertnishihara at this weekend, I will continue this pr. Need your assistance.

**Comment by richardliaw** [comment]
Hey @knightXun are you blocked on anything here?

**Comment by knightXun** [comment]
Do ray-project has slack or wechat group?

**Comment by richardliaw** [comment]
Thanks! No we don't have a slack. We have a mailing list (
ray-dev@googlegroups.com) if you're looking for a discussion space.

File: docker/kubernetes/ray/images/run.sh (hunk 2 above)
Comment (knightXun) [reply docker/kubernetes/ray/images/run.sh]: sorry, last week i was busy on kubeless.Now let's continue.
--redis-shard-ports specify the other redis servers but how can ray head use the redis server?

Comment (knightXun) [reply docker/kubernetes/ray/images/run.sh]: I plan to use HostNet mode for both ray head and redis. But this means redis and ray head must be in the same host.

Comment (knightXun) [reply docker/kubernetes/ray/images/run.sh]: Can I change some script in python/ray? I want one container have just one process.

**Comment by knightXun** [comment]
@richardliaw last 6 days.I was busy on kubeless. Maybe ray can use kubeless in the future.
In next week, I will spend my time on ray.
Here i have a question, maybe stupid: Can I change some script in python/ray? I want one container have just one process.

**Comment by robertnishihara** [comment]
For now, I think we should have one container correspond to one "Ray node". Otherwise it will be hard to take advantage of shared memory across workers on the same physical machine.

**Comment by knightXun** [comment]
@robertnishihara ok, let's discuss it in the future.

**Comment by knightXun** [comment]
I notice ray and some ML package are imported in the same *.py. Maybe I should write a script to build right image.  I also notice some import values such like gpu numbers. So I Must write them in the value.yaml @robertnishihara

**Comment by robertnishihara** [comment]
@knightXun, you should be able to ignore GPUs for now (Ray will try to determine the right number of GPUs when it is started).

**Comment by knightXun** [comment]
@robertnishihara i want know when autoscale ray is better. k8s can autoscale rc/deployment by cpu/memory usage. and does one node have merely one ray slave？

**Comment by robertnishihara** [comment]
Each docker container should run one full "ray node". As for austocaling, let's start by getting something working without autoscaling and then take it from there.

**Comment by knightXun** [comment]
@richardliaw @robertnishihara hi, I send a commit about helm.please review.

**Comment by knightXun** [comment]
jenkins retest this please

**Comment by knightXun** [comment]
@robertnishihara We can use configmap to mount ray code.

**Comment by knightXun** [comment]
@richardliaw  @robertnishihara  
It's maybe hard to provide a common kubernetes helm for every project.Can I provider a offical guidance for every user and rewrite some examples？ I believe it will be easy for me and make sense for users.Let us discuss common kubernetes-way in the future.

**Comment by virtualluke** [comment]
Why split redis? Sorry reading this late in the game.  We are using ray effectively in k8s, haven't done work on autoscaler yet (months overdue).  We haven't changed the yaml files much at all from Robert's initial ray on kubernetes page.

**Comment by knightXun** [comment]
@virtualluke spilit redis from ray-slave may be better for management, we can put all containers in a same pod, then do health-check for each containers, limit their memory/cpu usage even manage their logs.
All process in a single container is not a good way, had to do health-check. If one component is down, what we should do? restart the whole container that contains all containers?
As for autoscaler, i think the existing autoscaler is not based on kubernetes-way. If we want to use k8s, we must use k8s-way autoscaler.
That's what i do in huawei PaaS.

**Comment by robertnishihara** [comment]
Closing due to inactivity, note some recent work in #3126.

**Pull request closed without merging**